### PR TITLE
Optimize report delivery from client to active smart server in HA mode

### PR DESCRIPTION
**Pull request by philo-he**

### CODE REVIEW

File: smart-client/src/main/java/org/smartdata/client/SmartClient.java [hunk 1]
```diff
@@ -145,6 +159,38 @@ private void checkOpen() throws IOException {
...
+        fw.close();
+      }
+    }
+  }
+
+  /**
+   * Get recorded active server address.
+   */
+  private InetSocketAddress getActiveServerAddress() {
+    Scanner scanner = new Scanner(ACTIVE_SMART_SERVER_FILE_PATH);
```
Comment (lipppppp) [inline smart-client/src/main/java/org/smartdata/client/SmartClient.java]: Scanner(new File(ACTIVE_SMART_SERVER_FILE_PATH))

File: smart-client/src/main/java/org/smartdata/client/SmartClient.java [hunk 2]
```diff
@@ -145,6 +159,38 @@ private void checkOpen() throws IOException {
...
+    if (scanner.hasNextLine()) {
+      String address = scanner.nextLine();
+      String[] strings = address.split(":");
```
Comment (lipppppp) [inline smart-client/src/main/java/org/smartdata/client/SmartClient.java]: This address is like hostname/ip:port, and strings[0] is hostname/ip. It's not correct hostname for InetSocketAddress.

**Comment by philo-he** [comment]
I just added another solution. The report will be delivered to multiple smart servers simultaneously. And if one report task is successfully executed, other tasks will be canceled. This will be a default strategy. If user would like to fallback to previous one-by-one trying strategy, a property (see code) needs to be set to false in configuration on client side (HBase, Spark, etc).

**Comment by philo-he** [comment]
This new solution should also work well on windows environment. For one-by-one trying strategy with local file to maintain active server, windows support may take few effort. @lipppppp, you can try to do that support by yourself and then port it to your customer's windows environment.

**Pull request merged**